src/prisma/cli/_node.py

```diff
@@ -0,0 +1,320 @@
+        )
+
+
+class Strategy(ABC):
+    resolver: Literal['nodejs-bin', 'global', 'nodeenv']
+
+    # TODO: support more options
+    def run(
+        self,
+        *args: str,
+        check: bool = False,
+        cwd: Path | None = None,
+        stdout: File | None = None,
+        stderr: File | None = None,
+        env: Mapping[str, str] | None = None,
+    ) -> subprocess.CompletedProcess[bytes]:
+        """Call the underlying Node.js binary.
+
+        The interface for this function is very similar to `subprocess.run()`.
+        """
+        return self.__run__(
+            *args,
+            check=check,
+            cwd=cwd,
+            stdout=stdout,
+            stderr=stderr,
+            env=_update_path_env(
+                env=env,
+                target_bin=self.target_bin,
+            ),
+        )
+
+    @abstractmethod
+    def __run__(
+        self,
+        *args: str,
+        check: bool = False,
+        cwd: Path | None = None,
+        stdout: File | None = None,
+        stderr: File | None = None,
+        env: Mapping[str, str] | None = None,
+    ) -> subprocess.CompletedProcess[bytes]:
+        """Call the underlying Node.js binary.
+
+        This should not be directly accessed, the `run()` function should be used instead.
+        """
+
+    @property
+    @abstractmethod
+    def target_bin(self) -> Path:
+        """Property containing the location of the `bin` directory for the resolved node installation.
+
+        This is used to dynamically alter the `PATH` environment variable to give the appearance that Node
+        is installed globally on the machine as this is a requirement of Prisma's installation step, see this
+        comment for more context: https://github.com/RobertCraigie/prisma-client-py/pull/454#issuecomment-1280059779
+        """
+        ...
+
+
+class NodeBinaryStrategy(Strategy):
+    resolver: Literal['global', 'nodeenv']
+
+    def __init__(
+        self,
+        *,
+        path: Path,
+        resolver: Literal['global', 'nodeenv'],
+    ) -> None:
+        self.path = path
+        self.resolver = resolver
+
+    @property
+    def target_bin(self) -> Path:
+        return self.path.parent
+
+    def __run__(
+        self,
+        *args: str,
+        check: bool = False,
+        cwd: Path | None = None,
+        stdout: File | None = None,
+        stderr: File | None = None,
+        env: Mapping[str, str] | None = None,
+    ) -> subprocess.CompletedProcess[bytes]:
+        print('--- debug ---')
+        if platform.name() == 'windows':
+            for p in self.path.parent.rglob('*'):
+                if 'node_modules' in str(p):
+                    continue
+
+                print(p)
+        print('--- end ---')
+        print('-- list dir --')
+        for p in self.path.parent.iterdir():
+            print(p)
+        print('-- end --')
+
+        path = str(self.path.absolute())
+        log.debug('Executing binary at %s with args: %s', path, args)
```

Comment:
This is more an aside... Why this format over f-strings?

Reply:
It's so that we don't waste computing time on formatting the log message when debug logging isn't turned on